### DISCUSSION TITLE
Flatten the AST

### DISCUSSION THREAD

**Discussion by iagobaapellaniz** [kdheepak/Pandoc.jl#19] [Ideas]
I'm sorry if this is not the place to discuss this. These days I wanted to create some filters for my documents. The problem is that I need to "walk" all the AST in order to apply the filter. Typically so it happens.

Finally, I created the following functions (I write them here as they appear in my own code)
```julia
# In Pandoc.jl element properties are either
# - Element, Citation, Attr
# - Vector{Block}
# - Vector{Inline}
# - Vector{Vector{Block}}
# - Vector{Pair{Vector{Inline}, Vector{Block}}}
# or basic types as String, Int64, ...

function flatten(doc::Pandoc.Document) 
    vcat(flatten.(doc.blocks)...)
end

function flatten(el::Union{Pandoc.Element, Pandoc.Citation, Pandoc.Attr})
    o = Any[el]
    for prop in propertynames(el)
        elprop = getproperty(el, prop)
        t = typeof(elprop)
        if hasmethod(flatten, (t,))
            push!(o, flatten(elprop)...)
        end
    end
    o
end

function flatten(vec::Vector{<:Pandoc.Element})
    o = Any[]
    for el in vec
        push!(o, flatten(el)...)
    end
    o
end

function flatten(vec::Vector{Vector{<:Pandoc.Block}})
    o = Any[]
    for el in vec
        push!(o, flatten(el)...)
    end
    o
end

function flatten(vec::Vector{Pair{Vector{<:Pandoc.Inline}, Vector{<:Pandoc.Block}}})
    o = Any[]
    for pair in vec
        push!(o, flatten(pair[1])...)
        push!(o, flatten(pair[2])...)
    end
    o
end

function flattenby(type::Type, el)
    identity.(filter(x -> isa(x, type), flatten(el)))
end

function flattenby(cond::Function, el)
    identity.(filter(cond, flatten(el)))
end

function indexof(el, col)
    findfirst(isequal(el), col)
end

function getparent(el, col)
    for i in (indexof(el, col)-1):-1:1
        if in(el, flatten(col[i]))
            return col[i]
        end
    end
end
```

Whit these I get many things done. I can flatten the AST so finding certain elements is easier; I can obtain the index or position of them in the document (this is useful to determine what comes first); and finally, I am able to get the parent element. This is due to the fact that two elements are different even if they are equally constructed
```julia
julia> Pandoc.Div() == Pandoc.Div()
false
```
unless they are of the type of `Pandoc.Space()`, `Pandoc.HorizontalRule()` and similar.

On top of this, if someone edits now the contents of an element of the flattened vector, it is reflected on the original object. I think it mutates the object the way one would expect.

What do you think? Is it useful? I'm pretty sure that this is not the best way to implement this. Nevertheless, I wanted to share it.

I think Pandoc.jl could have a `Pandoc.DocumentAnnotated(str)` function which could return an annotated AST, where besides the type, content, and other pandoc fields and elements, one could access the element's index, parent element, and similar things as it happens on HTML DOM elements or similars.

Thanks you very much for the consideration.

Besides, I benchmarked the `flatten` function and it seems to be quite fast with few allocations (as many as elements in the document). Cheers.

**Comment by kdheepak** [comment]
Interesting approach with `flatten`!

I think `Pandoc.jl` should integrate with https://juliacollections.github.io/AbstractTrees.jl/stable/ and that way you'd be able to traverse the tree more easily using functionality from that package. Unfortunately I haven't gotten around to doing that yet.

I don't quite follow what you mean by `DocumentAnnotated` though. Can you elaborate with a more concrete example?

**Reply by iagobaapellaniz** [reply]
I can write some example based on the idea, but basically it would be to wrap every Pandoc type with an `Annotated{T}` type while `DocumentAnnotated(str)` parses the document.

The aim would be to store things like `parent` in the `Annotated{T}` wrapper. But, if the `AbstractTree` approach works, the idea makes no sense 😅. I did not know about `AbstractTree`.

**Comment by iagobaapellaniz** [comment]
For the moment I have the following ...
```julia
const SymOrInt = Union{Symbol, Int}

mutable struct Annotated{T}
    ast::T
    fast::Vector{Any}
    accesschain::Vector{Vector{SymOrInt}}
    function Annotated(ast)
        new{typeof(ast)}(ast, flatten(ast)...)
    end
end

const PandocTypes = Union{
        Pandoc.Attr,
        Pandoc.Caption,
        Pandoc.Cell,
        Pandoc.Citation,
        Pandoc.Row,
        Pandoc.ShortCaption,
        Pandoc.TableBody,
        Pandoc.TableFoot,
        Pandoc.TableHead,
        Pandoc.Element
    }

function flatten(doc::Pandoc.Document)
    f, a = flatten(doc.blocks, SymOrInt[:blocks])
    vcat([doc], f), vcat([SymOrInt[]], a)
end

function flatten(el::PandocTypes, accesschain::Vector{SymOrInt})
    vo = Any[]
    vl = Vector{SymOrInt}[]
    for propname in propertynames(el)
        prop = getproperty(el, propname)
        t = typeof(prop)
        if hasmethod(flatten, (t, Vector{SymOrInt}))
            o, l = flatten(prop, vcat(accesschain, [propname]))
            push!(vo, o...)
            push!(vl, l...)
        end
    end
    vcat([el], vo), vcat([accesschain], vl)
end

function flatten(vec::Union{Vector{<:Pandoc.Element}, Vector{Vector{Pandoc.Block}}}, accesschain::Vector{SymOrInt})
    vo = Any[]
    vl = Vector{SymOrInt}[]
    for i in axes(vec,1)
        o, l = flatten(vec[i], vcat(accesschain, [i]))
        push!(vo, o...)
        push!(vl, l...)
    end
    vo, vl
end

function flatten(vec::Vector{Pair{Vector{Pandoc.Inline}, Vector{Pandoc.Block}}}, accesschain::Vector{SymOrInt})
    vo = Any[]
    vl = Vector{SymOrInt}[]
    for i in axes(vec, 1)
        o2, l1 = flatten(vec[i][1], vcat(accesschain, [i, 1]))
        o2, l2 = flatten(vec[i][2], push!(accesschain, [i, 2]))
        push!(vo, o1...)
        push!(vo, o2...)
        push!(vl, l1...)
        push!(vl, l2...)
    end
    vo, vl
end

flatten(ast::Union{PandocTypes, Vector{<:Pandoc.Element}, Vector{Vector{Pandoc.Block}}, Vector{Pair{Vector{Pandoc.Inline}, Vector{Pandoc.Block}}}}) = flatten(ast, SymOrInt[])
```

This way, one wrappes the document (`ast`) and adds two more properties: `fast`(flattened AST) and `accesschain` (which stores the keychain to access any element in `fast`. From here getting children, parents, siblings, previous, next, etc should be trivial. 

If one mutates an object in `fast` the object in `ast` also gets updated (and also in the original AST):
```julia
julia> doc = Pandoc.Document(str);

julia> adoc = Annotated(doc);

julia> adoc.fast[3]
Pandoc.Str("Hello")

julia> adoc.fast[3].content = "Hola";

julia> ac = adoc.accesschain[3]
4-element Vector{Union{Int64, Symbol}}:
  :blocks
 1
  :content
 1

julia> doc.blocks[1].content[1]
Pandoc.Str("Hola")
```

Moreover the `Annotated{T}(ast)` constructor works with any of the types `flatten` does. Hence, it can be used for a subset of the document instead of the whole. 

And the benchmarks are quite nice IMHO
```julia
julia> @benchmark Pandoc.Document(str)
BenchmarkTools.Trial: 58 samples with 1 evaluation.
 Range (min … max):  75.214 ms … 92.562 ms  ┊ GC (min … max): 0.00% … 0.00%
 Time  (median):     87.953 ms              ┊ GC (median):    0.00%
 Time  (mean ± σ):   86.460 ms ±  4.551 ms  ┊ GC (mean ± σ):  0.05% ± 0.37%

                                          ▄   ▄█    ▂          
  ▄▁▁▄▆▄▄▆▁▁▁▁▁▄▄▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁█▄▆▁█▄█▆██▆▆▆▆█▆▄▁▄▄▁▁▄ ▁
  75.2 ms         Histogram: frequency by time        92.2 ms <

 Memory estimate: 1.17 MiB, allocs estimate: 16062.

julia> @benchmark Annotated(doc)
BenchmarkTools.Trial: 4690 samples with 1 evaluation.
 Range (min … max):  992.584 μs …   4.392 ms  ┊ GC (min … max): 0.00% … 73.51%
 Time  (median):       1.004 ms               ┊ GC (median):    0.00%
 Time  (mean ± σ):     1.065 ms ± 429.455 μs  ┊ GC (mean ± σ):  5.33% ±  9.87%

  █▁                                                             
  ██▆▁▃▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▅█ █
  993 μs        Histogram: log(frequency) by time        4.2 ms <

 Memory estimate: 827.36 KiB, allocs estimate: 12908.
```

**Reply by kdheepak** [reply]
That's awesome! Would you like to submit a PR? I'm not sure when I'll get to implementing AbstractTrees but I think this could exist in parallel regardless.

**Reply by iagobaapellaniz** [reply]
Certainly I will if you agree on implementing this feature! I have this code in a side project of mine, but I could export it to `Pandoc.jl`. Thanks for the consideration

**Reply by iagobaapellaniz** [reply]
After investigating `AbstractTrees.jl`, I decided to give it a try. Nevertheless, I learned a lot with my approach. Thanks! I'll update you about the possible PR.

**Reply by kdheepak** [reply]
Thanks for considering making a PR! Let me know if I can help with anything.

**Comment by iagobaapellaniz** [comment]
Hey @kdheepak!

I have prepared a branch https://github.com/iagobaapellaniz/Pandoc.jl/tree/abstract-trees-interface which works with `AbstractTrees`.

Could you create a new branch on the main repository, maybe `iagobaapellaniz-nodes`, such that I create a PR against that and not against the `main` branch?

**Reply by kdheepak** [reply]
I made a new branch here: https://github.com/kdheepak/Pandoc.jl/tree/iagobaapellaniz-nodes

You can submit a PR to `main` and that's fine too if you want.
